File: src/main/java/com/example/smarthouse/sensor/smoke/SmokeDetectorSensor.java
package com.example.smarthouse.sensor.smoke;

import com.example.smarthouse.sensor.DefaultSensor;
import com.example.smarthouse.sensor.config.SensorProtocol;
import com.example.smarthouse.sensor.config.SensorState;
import com.example.smarthouse.sensor.config.SensorType;
import com.example.smarthouse.sensor.message.MessageType;
import com.example.smarthouse.sensor.message.SensorMessage;
import lombok.Data;

import java.util.Date;

@Data
public class SmokeDetectorSensor extends DefaultSensor {
    private boolean smokeDetected;

    public SmokeDetectorSensor(int sensorId, String sensorName, SensorProtocol sensorProtocol) {
        super(sensorId, sensorName, SensorState.ON, sensorProtocol, SensorType.SMOKE_DETECTOR_SENSOR);
        this.smokeDetected = false;
    }

    public void detectSmoke() {
        smokeDetected = true;
        System.out.println("Smoke detected");
        activateAlarm(); // Activate alarm when smoke is detected
    }

    public void clearSmoke() {
        smokeDetected = false;
        System.out.println("Smoke cleared");
    }

    private void activateAlarm() {
     // Calling an alarm from commands
    }

    @Override
    public SensorMessage getDataFromSensor() {
        return new SensorMessage(
                getSensorId(),
                smokeDetected ? 1 : 0,
                new Date(),
                MessageType.SENSOR_DATA,
                smokeDetected ? "Smoke detected" : "No smoke",
                getSensorType()
        );
    }
}

File: src/main/kotlin/com/example/sensor/SmokeDetectorSensor.kt
package com.example.smarthouse.sensor.smoke

import com.example.config.SensorProtocol
import com.example.config.SensorState
import com.example.config.SensorType
import com.example.message.MessageType
import com.example.message.Message
import com.example.smarthouse.sensor.DefaultSensor

import java.util.*

data class SmokeDetectorSensor(val sensorId: Int, val sensorName: String?, val sensorProtocol: SensorProtocol?) :
    DefaultSensor(sensorId, sensorName, SensorState.ON, sensorProtocol, SensorType.SMOKE_DETECTOR_SENSOR) {
    private var smokeDetected = false
    fun detectSmoke() {
        smokeDetected = true
        println("Smoke detected")
        activateAlarm() // Activate alarm when smoke is detected
    }

    fun clearSmoke() {
        smokeDetected = false
        println("Smoke cleared")
    }

    private fun activateAlarm() {
        // Calling an alarm from commands
    }

    override val dataFromSensor: Message
        get() = Message(
            getSensorId(),
            if (smokeDetected) 1 else 0,
            Date(),
            MessageType.SENSOR_DATA,
            if (smokeDetected) "Smoke detected" else "No smoke",
            getSensorType()
        )
}
